[ArgPromotion] Handle pointer arguments of recursive calls

vedantparanjape-amd opened the pull request:
Argument promotion doesn't handle recursive function calls to promote arguments. This patch adds functionality to handle self recursive function calls, i.e. whose SCC size is 1. Due to complexity of ValueTracking in recursive calls with SCC size greater than 1, we bail out in such cases

Fixes #1259

vedantparanjape-amd committed:
[ArgPromotion] Handle pointer arguments of recursive calls

Argument promotion doesn't handle recursive function calls to promote
arguments. This patch adds functionality to handle self recursive
function calls, i.e. whose SCC size is 1. Due to complexity of Value
Tracking in recursive calls with SCC size greater than 1, we bail out
in such cases.

jdoerfert commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -609,13 +609,66 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
+    if (IsRecursive && CI && CI->getFunction() == Arg->getParent()) {
+      dbgs() << "Found recursive call\n";
+      Type *Ty = CI->getType();
+      Value *Ptr = CI->getArgOperand(U->getOperandNo());
```
Isn't this just `*U`?

jdoerfert commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -609,13 +609,66 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
 
+    auto *CI = dyn_cast<CallInst>(V);
+    if (IsRecursive && CI && CI->getFunction() == Arg->getParent()) {
+      dbgs() << "Found recursive call\n";
```
Leftover

jdoerfert commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -609,13 +609,66 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
     }
 
+    auto *CI = dyn_cast<CallInst>(V);
+    if (IsRecursive && CI && CI->getFunction() == Arg->getParent()) {
```
How can this be a different function?

jdoerfert reviewed (commented):
I haven't looked at this code in a long time. Left some comments though.

Tests missing.

arsenm reviewed (commented):
Reverse ping

vedantparanjape-amd commented:
@arsenm @jdoerfert I have addressed the comment, but I am stuck at one thing. The testcase: Transforms/ArgumentPromotion/pr42028-recursion.ll

```
A -> B -> C
     ↑____|
```

It has a SCC like this.

arsenm commented:
> @arsenm @jdoerfert I have addressed the comment, but I am stuck at one thing. The testcase: Transforms/ArgumentPromotion/pr42028-recursion.ll
> 
> ```
> A -> B -> C
>      ↑____|
> ```
> 
> It has a SCC like this. Somehow, not able to handle the case with call instructions.

Handle in what way? It breaks, or is sub-optimal? If it's not broken, it doesn't need to be perfect in the first attempt

vedantparanjape-amd commented:
> > @arsenm @jdoerfert I have addressed the comment, but I am stuck at one thing. The testcase: Transforms/ArgumentPromotion/pr42028-recursion.ll
> > ```
> > A -> B -> C
> >      ↑____|
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It has a SCC like this. Somehow, not able to handle the case with call instructions.
> 
> Handle in what way? It breaks, or is sub-optimal? If it's not broken, it doesn't need to be perfect in the first attempt

It breaks (SE), while handling that case, if we try to do argument promotion for such a case it will lead to recursion.

arsenm commented:
> @arsenm @jdoerfert I have addressed the comment,

This PR hasn't been updated?

vedantparanjape-amd commented:
> > @arsenm @jdoerfert I have addressed the comment,
> 
> This PR hasn't been updated?

Just a min, cleaning up and pushing.

vedantparanjape-amd commented:
> > @arsenm @jdoerfert I have addressed the comment,
> 
> This PR hasn't been updated?

Done, please take a look! adding the testcases in sometime.

arsenm commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -610,13 +610,59 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
       // unknown users
     }
 
+    if (IsSelfRecursive && isa<CallBase>(V)) {
+      Value *Ptr = dyn_cast<Value>(U);
```
unchecked dyn_cast, which is also redundant with the isa. Just dyn_cast to CallBase in the first place? You also need to check this is actually the callee argument

vedantparanjape-amd replied:
U is the Use, and not the User, so it's not redundant with the isa in the if condition. Am I missing something ?

vedantparanjape-amd replied:
> You also need to check this is actually the callee argument

That's being checked below, right ?

```c
      if (PtrArg != Arg)
        return false;
```

arsenm commented:
> Done, please take a look! adding the testcases in sometime.

The test is the most important part of the review

vedantparanjape-amd commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -744,8 +794,10 @@ static Function *promoteArguments(Function *F, FunctionAnalysisManager &FAM,
+    if (CB->getFunction() == F) {
       IsRecursive = true;
+      IsSelfRecursive = true;
+    }
```
norecurse
This function attribute indicates that the function does not call itself either directly or indirectly down any possible call path. This produces undefined behavior at runtime if the function ever does recurse.

Yes, but is it guaranteed that each such function will be tagged norecursive ? Also, we'd still need to check for self recursion. That said, can this be submitted as a separate patch ?

vedantparanjape-amd commented:
> > Done, please take a look! adding the testcases in sometime.
> 
> The test is the most important part of the review

I have added the testcase as well, and addressed the comments best I could.

vedantparanjape-amd committed:
Add a check to stop recursive promotion of ptr args

vedantparanjape-amd commented:
@arsenm implemented a check on function call args to see if they are being dereferenced somewhere above the call. If they are not we can bail out, and not promote the argument. This will remove the need for IsSelfRecursive flag, and provide a better way to detect recursive arg promotions.

arsenm commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -610,13 +610,78 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
     }
 
+    auto *CB = dyn_cast<CallBase>(V);
+    Value *PtrArg = dyn_cast<Value>(U);
```
Why do you need to dyn_cast to Value?

vedantparanjape-amd replied:
right, even cast will be enough.

arsenm commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -610,13 +610,78 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
   }
 
-  if (NeededDerefBytes || NeededAlign > 1) {
+  // Incase of functions with recursive calls, this check will fail when it
```
"In case"

arsenm commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -610,13 +610,78 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
+    auto *CB = dyn_cast<CallBase>(V);
+    Value *PtrArg = dyn_cast<Value>(U);
+    if (IsRecursive && CB && PtrArg) {
+      Type *PtrTy = PtrArg->getType();
```
This looks a lot like reinventing HandleEndUser

vedantparanjape-amd replied:
Yes, but for a first try it seems okay to do. It's not exactly same as HandleEndUser. Mainly the issue is that some checks in that functor like (below) fail for CallIBase objects.

```cpp
    TypeSize Size = DL.getTypeStoreSize(Ty);
    // Don't try to promote scalable types.
    if (Size.isScalable())
      return false;
```

vedantparanjape-amd commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -610,13 +610,78 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
+      // testcases.
+      bool doesPointerResolve = false;
+      for (auto Load : Loads)
+        if (Load->getPointerOperand() == PtrArg &&
+            !Load->getType()->isPointerTy())
+          doesPointerResolve = true;
```
Well, to be fair. Even HandleEndUser is doing that. It is checking if the Load type is a pointer.

> You can't rely on the type of the load to assume it can't load the pointer. 

That is a relatively corner case, not being handled anywhere in the non recursive code.

vedantparanjape-amd replied:
it is a mechanical pass check. The crux here is, if some ptr argument isn't loaded into a value then we shouldn't try to promote it, as due to opaque pointers we don't the depth of the pointers.

vedantparanjape-amd committed:
Address review comments

vedantparanjape-amd commented:
sorry for the code review requests, I believe I messed up while rebasing my local branch.

efriedma-quic commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -610,13 +610,78 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
+  //   %resbar = call i32 @fun(ptr %x)
+  //   ...
+  // }
+  if (!IsRecursive && (NeededDerefBytes || NeededAlign > 1)) {
```
This doesn't look right.  You can't skip this check just because the function is recursive; you need to specifically skip the recursive calls you processed earlier.

vedantparanjape-amd replied:
Okay, more like set a flag ONLY when we successfully promote some recursive argument and then only skip this check based on this flag ? seems nice, will do this.

vedantparanjape-amd replied:
Ah by bad, it's not that simple.

efriedma-quic commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -610,13 +610,78 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
+      Type *PtrTy = PtrArg->getType();
+      Align PtrAlign = PtrArg->getPointerAlignment(DL);
+      APInt Offset(DL.getIndexTypeSizeInBits(PtrArg->getType()), 0);
+      PtrArg = PtrArg->stripAndAccumulateConstantOffsets(
```
If the offset isn't zero, there's no way to make the recursion work correctly in general: the overall offset is multiplied by the number of levels of recursion, and you can only promote a finite number of bytes.

Given that, we don't need to track the call in ArgParts at all.  You just need to check that the offset is zero, and that we're passing the pointer to the corresponding argument of the call.

vedantparanjape-amd replied:
Done

vedantparanjape-amd committed:
Address review 2 comments

vedantparanjape-amd commented:
@efriedma-quic please take a look, updated my code addressing the review!

vedantparanjape-amd committed:
minor nitpick

vedantparanjape-amd commented:
ping @nikic if you can review this part of code, as it seems you have authored this part, thanks!

efriedma-quic commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -610,15 +614,80 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
+      int64_t Off = Offset.getSExtValue();
+      if (Off)
+        LLVM_DEBUG(dbgs() << "ArgPromotion of " << *Arg << " failed: "
+                          << "pointer offset is not equal to zero\n");
```
Did you mean to actually do something when you see a non-zero offset?  This just prints.

Also, you need to implement the check that the argument number matches

vedantparanjape-amd replied:
> Also, you need to implement the check that the argument number matches

Sorry I don't get this. I am doing that check above on L632 already. I am checking if the Use of the said argument matches the argument. So for a recursive call this will be true, no ? I can think of a case like this, in this case there will be two uses of ptr %a if I am not wrong, so the handler will be called twice anyways!

```c
def fun(ptr %a, ptr %b) {
    ...
    ...
    call @func(ptr %a, ptr %a)
    ...
}
```

efriedma-quic replied:
Say you have something like:

```
define void func(ptr %a, ptr %b) {
    ...
    ...
    call @func(ptr %b, ptr %a)
    ...
}
```

In this case, you can't promote %a on its own, because %a and %b are intertwined.  So we need to bail out.  (In theory, you could handle some subset of cases like this, but let's leave that for a different patch.)

I don't see how that's related to the "doesPointerResolve" check, which... isn't actually a correctness check in that sense.  It's there to ensure the optimization itself terminates.

vedantparanjape-amd replied:
Ah, this makes sense now. I will implement this.

vedantparanjape-amd replied:
> Say you have something like:
> 
> ```
> define void func(ptr %a, ptr %b) {
>     ...
>     ...
>     call @func(ptr %b, ptr %a)
>     ...
> }
> ```
> 
> In this case, you can't promote %a on its own, because %a and %b are intertwined. So we need to bail out. (In theory, you could handle some subset of cases like this, but let's leave that for a different patch.)
> 
> I don't see how that's related to the "doesPointerResolve" check, which... isn't actually a correctness check in that sense. It's there to ensure the optimization itself terminates.

I implemented this, it was rather simple. lmk if it seems correct.

efriedma-quic replied:
Looks right.

efriedma-quic commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -422,13 +422,16 @@ doPromotion(Function *F, FunctionAnalysisManager &FAM,
   APInt Bytes(64, NeededDerefBytes);
 
+  if (RecursiveCalls.size())
+    return true;
```
Again, you can't just check that there's a recursive call; even if there's a recursive call, the checks in this function are relevant for the other calls.

vedantparanjape-amd replied:
I don't think there is an easy way to implement this. I have pushed another patch but it fails even for the current testcase, but the workaround for it creates issues with other testcases too.

problem is that it looks at the recursive function's callee in the attached testcase and it fails at `isDereferenceableAndAlignedPointer` check for that call. So, I am not sure then if the attached testcase is correct in the first place ?

efriedma-quic replied:
Reading the testcase... the argument isn't marked dereferenceable for the definition of foo(), and we don't have any information in bar(), so the current logic can't prove it's dereferenceable.

It might be possible to prove the argument is dereferenceable because there's a load on all paths through foo(), but that logic shouldn't be in ArgPromotion; it should be in function-attrs, or something like that.

vedantparanjape-amd replied:
> Reading the testcase... the argument isn't marked dereferenceable for the definition of foo(), and we don't have any information in bar(), so the current logic can't prove it's dereferenceable.

Yes, that's true. Do, you suggest I add the !dereferenceable metadata to this and run the testcase ?

> It might be possible to prove the argument is dereferenceable because there's a load on all paths through foo(), but that logic shouldn't be in ArgPromotion; it should be in function-attrs, or something like that.

Makes sense

efriedma-quic replied:
> Do, you suggest I add the !dereferenceable metadata to this and run the testcase ?

I was thinking a dereferenceable attribute, not metadata.  But yes, that probably makes sense.

> can this be a part of a separate patch

Sure.

vedantparanjape-amd replied:
> > Do, you suggest I add the !dereferenceable metadata to this and run the testcase ?
> 
> I was thinking a dereferenceable attribute, not metadata. But yes, that probably makes sense.

```
define internal i32 @foo(ptr dereferenceable(16) %x, i32 %n, i32 %m) {
entry:
  %tmp = icmp ne i32 %n, 0
  br i1 %tmp, label %cond_true, label %cond_false

cond_true:                                        ; preds = %entry
  %tmp2 = load i32, ptr %x, align 4
  br label %return

cond_false:                                       ; preds = %entry
  %tmp5 = load i32, ptr %x, align 4
  %tmp7 = sub i32 %n, 1
  %tmp9 = call i32 @foo(ptr %x, i32 %tmp7, i32 %tmp5)
  %tmp11 = sub i32 %n, 2
  %tmp14 = call i32 @foo(ptr %x, i32 %tmp11, i32 %m)
  %tmp15 = add i32 %tmp9, %tmp14
  br label %return

cond_next:                                        ; No predecessors!
  br label %return

return:                                           ; preds = %cond_next, %cond_false, %cond_true
  %retval.0 = phi i32 [ %tmp2, %cond_true ], [ %tmp15, %cond_false ], [ undef, %cond_next ]
  ret i32 %retval.0
}

define i32 @bar(ptr %x, i32 %n, i32 %m) {
entry:
  %load0 = load i32, ptr %x, align 4
  %tmp3 = call i32 @foo(ptr %x, i32 %n, i32 %m)
  br label %return

return:                                           ; preds = %entry
  ret i32 %tmp3
}
``` 
I tried something like this, doesn't seem to work either! Can you take a look what might be wrong ?

> > can this be a part of a separate patch
> 
> Sure.

efriedma-quic replied:
The code currently only looks for dereferenceable attributes on the caller (i.e. bar()), not the function itself.

vedantparanjape-amd replied:
> The code currently only looks for dereferenceable attributes on the caller (i.e. bar()), not the function itself.

Even this doesn't seem to work or manually adding a load i32 on ptr %x

```
define i32 @bar(ptr dereferenceable(32) %x, i32 %n, i32 %m) {
entry:
  %tmp3 = call i32 @foo(ptr %x, i32 %n, i32 %m)
  br label %return

return:                                           ; preds = %entry
  ret i32 %tmp3
}
```

vedantparanjape-amd replied:
ping @efriedma-quic

efriedma-quic replied:
You might need to add an align attribute?  Otherwise, I'd suggest using a debugger to step into the dereferenceable check to see why it's failing.

vedantparanjape-amd replied:
@efriedma-quic sorry for the tag, but I was able to get it working with align. Pushed a cleaned up patch, hope it is in a mergeable state now :) and can get a LGTM!

vedantparanjape-amd committed:
Address review 3 comments

vedantparanjape-amd committed:
Address review 4 comments

arsenm commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -610,15 +621,89 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
+      // Transforms/ArgumentPromotion/2008-09-08-CGUpdateSelfEdge.ll
+      // testcases.
+      bool doesPointerResolve = false;
+      for (auto Load : Loads)
+        if (Load->getPointerOperand() == PtrArg &&
+            !Load->getType()->isPointerTy())
+          doesPointerResolve = true;
+
+      if (!doesPointerResolve)
+        return false;
```
Can you move this loop/assign bool/post-loop check to a predicate function with a comment? It's not obvious to me what "does pointer resolve" means

arsenm commented on llvm/test/Transforms/ArgumentPromotion/argpromotion-recursion-pr1259.ll:
```diff
@@ -0,0 +1,65 @@
+  br label %return
+
+return:                                           ; preds = %cond_next, %cond_false, %cond_true
+  %retval.0 = phi i32 [ %val, %cond_true ], [ %cmp2, %cond_false ], [ undef, %cond_next ]
```
Use poison instead of undef

arsenm commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -437,6 +440,13 @@ static bool allCallersPassValidPointerForArgument(Argument *Arg,
+    if (RecursiveCalls.contains(&CB))
+      return true;
+
+    // if (RecursiveCalls.size() &&
+    //     CB.getCalledFunction()->getName() == Callee->getName())
+    //   return true;
```
Commented out code

vedantparanjape-amd replied:
Yes, I will remove it. Was referring to it, in my thread with efriedma!

arsenm commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -422,13 +422,16 @@ doPromotion(Function *F, FunctionAnalysisManager &FAM,
   const DataLayout &DL = Callee->getParent()->getDataLayout();
   APInt Bytes(64, NeededDerefBytes);
 
+  // if (RecursiveCalls.size())
+  //   return true;
```
Commented out code

vedantparanjape-amd committed:
Refactor and add comments, also address review comments

efriedma-quic commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -610,6 +658,48 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
 
+    auto *CB = dyn_cast<CallBase>(V);
+    Value *PtrArg = cast<Value>(U);
+    if (IsRecursive && CB && PtrArg) {
```
This IsRecursive check doesn't look right: it checks whether the function contains a recursive call somewhere, but it doesn't check whether this particular call is recursive.

vedantparanjape-amd replied:
I had added a check to check if a function is self recursive, and not a chain of recursion (pr42028-recursion.ll). Does that seem okay?

vedantparanjape-amd replied:
pushed a patch addressing both the comments.

efriedma-quic replied:
You're still not checking if the call in question is recursive; a function can contain both recursive and non-recursive calls.

vedantparanjape-amd replied:
> You're still not checking if the call in question is recursive; a function can contain both recursive and non-recursive calls.

Oh so, just a check if CB.getCalledFunction() == CB.getFunction()! Done and pushed.

efriedma-quic replied:
Yes, that looks right.  (And with that, you don't really need the IsSelfRecursive check here.)

vedantparanjape-amd replied:
Should I still pass IsSelfRecursive ? seems it might come in handy, because IsRecursive is quite generic, as it just uses SCC.size()

efriedma-quic commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -420,11 +420,31 @@ doPromotion(Function *F, FunctionAnalysisManager &FAM,
 
+/// Returns true if the Ptr is loaded by any Load in the vector of
+/// Loads, and if the Loaded value is not a pointer.
+static bool checkIfPointerIsDereferenced(SmallVector<LoadInst *, 16> &Loads,
```
With the most recent changes, please check again if checkIfPointerIsDereferenced(); I suspect the other checks make this unnecessary.

vedantparanjape-amd replied:
This is still needed to prevent the failure in (pr42028-recursion.ll) case, as it is not self recursive. But, if I add the self recursive check I mentioned above, then it would be okay.

efriedma-quic commented on llvm/test/Transforms/ArgumentPromotion/argpromotion-recursion-pr1259.ll:
```diff
@@ -0,0 +1,65 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
```
This needs a lot more tests.  In particular, tests for the various cases we discussed that should not be transformed.

vedantparanjape-amd replied:
makes sense, I will add testcase for these two cases:

1) check for argument position of the caller and callee.
2) check if the pointer offset is non-zero

lmk, if I need to add anything more.

vedantparanjape-amd replied:
added two more testcases, not sure what more I can add, any suggestions ?

efriedma-quic replied:
You can probably put all the testcases in one file.

In terms of what to test... non-zero offsets, functions containing both recursive and non-recursive calls, a struct with multiple elements.

vedantparanjape-amd replied:
> a struct with multiple elements.

can you help with this ? not sure what is expected.

vedantparanjape-amd committed:
Add check for self recursion

vedantparanjape-amd committed:
Add check if call is recursion

vedantparanjape-amd committed:
remove isSelfRecursive check

vedantparanjape-amd committed:
Add check for argument number as well in recursive calls

vedantparanjape-amd committed:
Added more testcases

vedantparanjape-amd committed:
Add more tests, and move them to a folder

vedantparanjape-amd commented:
@efriedma-quic I have added all the testcases except for "struct with multiple elements" because I was not sure what to add in that case.

vedantparanjape-amd commented:
ping @efriedma-quic !

efriedma-quic commented:
Please put comments in each of the test files briefly explaining what you're looking for.

For "multiple elements", I mean aggregate promotion, like llvm/test/Transforms/ArgumentPromotion/aggregate-promote.ll

vedantparanjape-amd committed:
Added testcase for multiple elements

vedantparanjape-amd commented:
ping @efriedma-quic I have finished all the to-dos after adding the multiple element access testcase, can I get a LGTM ?

vedantparanjape-amd commented:
ping @efriedma-quic !

vedantparanjape-amd commented:
not sure if I am missing something, if yes, please let me know!

efriedma-quic commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -446,6 +473,7 @@ static bool allCallersPassValidPointerForArgument(Argument *Arg,
 /// parts it can be promoted into.
 static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
                          unsigned MaxElements, bool IsRecursive,
+                         bool isSelfRecursive,
```
isSelfRecursive is unused; you can delete it.

efriedma-quic reviewed (approved):
LGTM with one minor comment.

(In the future, please give a little more time for people to reply; I get a lot of review email every day.)

nikic commented:
Can you please update the PR description? It doesn't explain at all what (and why) the PR does and just mentions that it causes a compiler crash -- which it presumably no longer does?

nikic commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -422,9 +422,9 @@ doPromotion(Function *F, FunctionAnalysisManager &FAM,
-                                                  Align NeededAlign,
-                                                  uint64_t NeededDerefBytes) {
+static bool allCallersPassValidPointerForArgument(
+    Argument *Arg, SmallPtrSet<CallBase *, 4> &RecursiveCalls,
```
Use SmallPtrSetImpl instead.

nikic commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -610,6 +639,45 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
+        return false;
+
+      int64_t Off = Offset.getSExtValue();
+      if (Off) {
```
Calling stripAndAccumulateConstantOffsets() just to check for a zero offset afterwards is pretty pointless. Use stripPointerCasts() instead -- or just don't strip at all, now that we have opaque pointers.

vedantparanjape-amd replied:
I get your point that strips are anyways not there in opaque ptrs, but how to collect offsets ? I couldn't find a function which will accumulate constant offsets on a Value! Please take a look at testcase recursion-non-zero.ll submitted with this PR. We are trying to check if a ptr is being offset at some non-zero offset.

vedantparanjape-amd replied:
How do you propose I check zero offset ? because checking zero offset is a necessary check and it it's not we need to bail out in this optimization.

stripAndAccumulateConstantOffsets() doesn't do much, it is just traversing the GEP tree and appending the constant offsets to Offset variable.

nikic replied:
You check for zero offset using `PtrArg == Arg`.

vedantparanjape-amd replied:
@nikic the function https://github.com/llvm/llvm-project/blob/e4163c0927772f2ec73cf16d53e000614c419c45/llvm/lib/IR/Value.cpp#L731 pretty much runs a do-while loop and traverses the GEP tree to accumulate the offsets. If it encounters a bitcast (which it won't), it doesn't do much. So, I would want to ideally pull out just the traversal code out from the function, skipping the bitcasts code. But I think that can another patch, is it blocking this one ?

vedantparanjape-amd commented:
> LGTM with one minor comment.
> 
> (In the future, please give a little more time for people to reply; I get a lot of review email every day.)

Thanks, sorry for too many pings.

vedantparanjape-amd commented:
> Can you please update the PR description? It doesn't explain at all what (and why) the PR does and just mentions that it causes a compiler crash -- which it presumably no longer does?

Yes, I will do this.

vedantparanjape-amd committed:
Removed IsSelfRecursive from findArgParts

vedantparanjape-amd committed:
Use SmallPtrSetImpl in a helper function

vedantparanjape-amd committed:
minor fix to SmallPtrSetImpl

nikic commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -610,6 +638,45 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
+    Value *PtrArg = cast<Value>(U);
+    if (CB && PtrArg && CB->getCalledFunction() == CB->getFunction()) {
+      Type *PtrTy = PtrArg->getType();
+      APInt Offset(DL.getIndexTypeSizeInBits(PtrTy), 0);
+      PtrArg = PtrArg->stripAndAccumulateConstantOffsets(
+          DL, Offset,
+          /* AllowNonInbounds= */ true);
+      if (PtrArg != Arg)
+        return false;
+
+      if (Offset.getSignificantBits() >= 64)
+        return false;
+
+      int64_t Off = Offset.getSExtValue();
+      if (Off) {
+        LLVM_DEBUG(dbgs() << "ArgPromotion of " << *Arg << " failed: "
+                          << "pointer offset is not equal to zero\n");
+        return false;
+      }
```
Let me spell this out explicitly. Here is the change you should do:
```suggestion
      if (PtrArg != Arg)
        return false;
```

vedantparanjape-amd committed:
Remove use of stripAndAccumulateConstantOffsets

vedantparanjape-amd commented:
@nikic I have addressed all the requested changes, can you give a LGTM! thanks.

nikic commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -611,6 +639,33 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
     }
 
+    auto *CB = dyn_cast<CallBase>(V);
+    Value *PtrArg = cast<Value>(U);
```
Writing this as `cast<Value>(U)` works, but is pretty weird. `U.get()` would be better.

nikic commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -611,6 +639,33 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
 
+    auto *CB = dyn_cast<CallBase>(V);
+    Value *PtrArg = cast<Value>(U);
+    if (CB && PtrArg && CB->getCalledFunction() == CB->getFunction()) {
```
The `PtrArg` check here is unnecessary, it cannot be null.

nikic commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -611,6 +639,33 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
+      }
+
+      unsigned int ArgNo = Arg->getArgNo();
+      if (CB->getArgOperand(ArgNo) != Arg || U->getOperandNo() != ArgNo) {
```
The `CB->getArgOperand(ArgNo) != Arg`  check is redundant, `U->getOperandNo() != ArgNo` already does this more correctly.

nikic commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -611,6 +639,33 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
 
+    auto *CB = dyn_cast<CallBase>(V);
+    Value *PtrArg = cast<Value>(U);
+    if (CB && PtrArg && CB->getCalledFunction() == CB->getFunction()) {
```
This should probably also check that the function type of the call and function match? I wouldn't want to reason about how the arguments map if this is not the case.

vedantparanjape-amd replied:
What do you mean by function type, is it the thing returned by getFunctionType() ? I am not able to understand why it would be different ?

arsenm replied:
The call and the function have separate types that do not match:

```
declare float @ret_float()

define i32 @call_as_i32() {
  %val = call i32 @ret_float()
  ret i32 %val
}
```

vedantparanjape-amd replied:
created a new PR to address these changes, please take a look (https://github.com/llvm/llvm-project/pull/98657)

nikic commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -700,6 +755,10 @@ static bool areTypesABICompatible(ArrayRef<Type *> Types, const Function &F,
+  // Due to complexity of handling cases where the SCC has more than one
+  // component. We want to limit argument promotion of recursive calls to
+  // just functions that directly call themselves.
+  bool IsSelfRecursive = false;
```
This dead variable still hasn't been dropped...
